ngx_http_auth_digest_module.c

```diff
@@ -12,1499 +12,1576 @@
-          p[0] = '\0';
-          passwd_line.len = i - begin;
-          rc = ngx_http_auth_digest_verify_user(r, auth_fields, &passwd_line);
-
-          if (rc == NGX_HTTP_AUTH_DIGEST_USERNOTFOUND) {
-            rc = NGX_DECLINED;
-          }
-
-          if (rc != NGX_DECLINED) {
-            ngx_http_auth_digest_close(&file);
-            ngx_http_auth_digest_evasion_tracking(
-                r, alcf, NGX_HTTP_AUTH_DIGEST_STATUS_SUCCESS);
-            return rc;
-          }
-        }
-        idle = 1;
-        begin = i;
-      } else if (idle) {
-        idle = 0;
-        begin = i;
-      }
-    }
-
-    if (!idle) {
-      tail = n + tail - begin;
-      if (n == 0 && tail > 36) {
-        p = ngx_cpymem(passwd_line.data, &buf[begin], tail);
-        p[0] = '\0';
-        passwd_line.len = i - begin;
-        rc = ngx_http_auth_digest_verify_user(r, auth_fields, &passwd_line);
-        if (rc == NGX_HTTP_AUTH_DIGEST_USERNOTFOUND) {
-          rc = NGX_DECLINED;
-        }
-        if (rc != NGX_DECLINED) {
-          ngx_http_auth_digest_close(&file);
-          ngx_http_auth_digest_evasion_tracking(
-              r, alcf, NGX_HTTP_AUTH_DIGEST_STATUS_SUCCESS);
-          return rc;
-        }
-      } else {
-        ngx_memmove(buf, &buf[begin], tail);
-      }
-    }
-
-    if (n == 0) {
-      break;
-    }
-
-    offset += n;
-  }
-
-  ngx_http_auth_digest_close(&file);
-
-  // log only wrong username/password,
-  // not expired hash
-  int nc = ngx_hextoi(auth_fields->nc.data, auth_fields->nc.len);
-  if (nc == 1) {
-    ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
-                  "invalid username or password for %*s",
-                  auth_fields->username.len, auth_fields->username.data);
-  }
-
-  ngx_http_auth_digest_evasion_tracking(r, alcf,
-                                        NGX_HTTP_AUTH_DIGEST_STATUS_FAILURE);
-
-  // since no match was found based on the fields in the authorization header,
-  // send a new challenge and let the client retry
-  return ngx_http_auth_digest_send_challenge(r, &realm, auth_fields->stale);
+static ngx_int_t ngx_http_auth_digest_handler(ngx_http_request_t* r) {
+	off_t offset;
+	ssize_t n;
+	ngx_fd_t fd;
+	ngx_int_t rc;
+	ngx_err_t err;
+	ngx_str_t user_file, passwd_line, realm;
+	ngx_file_t file;
+	ngx_uint_t i, begin, tail, idle;
+	ngx_http_auth_digest_loc_conf_t* alcf;
+	ngx_http_auth_digest_cred_t* auth_fields;
+	u_char buf[NGX_HTTP_AUTH_DIGEST_BUF_SIZE];
+	u_char line[NGX_HTTP_AUTH_DIGEST_BUF_SIZE];
+	u_char* p;
+
+	if (r->internal) {
+		return NGX_DECLINED;
+	}
+
+	// if digest auth is disabled for this location, bail out immediately
+	alcf = ngx_http_get_module_loc_conf(r, ngx_http_auth_digest_module);
+
+	const int isLocalHost = ngx_http_auth_digest_is_loopback(&(r->headers_in.server)) == 0;
+
+	if (isLocalHost) {
+		const int isLocalHostBypass = ngx_strcmp(alcf->allow_localhost.data, "on") == 0;
+		if (isLocalHostBypass) {
+			return NGX_DECLINED;
+		}
+	}
```

Comment:
The `isLocalHostBypass` check is much faster and will most of the time fail since the default is `off`.
1. Can we convert `alcf->allow_localhost` to an `int` (different variable) in `ngx_http_auth_digest_merge_loc_conf` so we just have to compare 1 int instead of a string each request?
2. Its better to do this test first before we do the more expensive `isLocalHost` test.